### ISSUE TITLE
`torch.lu` CUDA gives different results for batched and unbatched cases

### ISSUE THREAD

**Issue by kurtamohler** [pytorch/pytorch#56312]
## 🐛 Bug

When `torch.lu` is given a batch of CUDA matrices, it gives different results than if we call `torch.lu` on each individual matrix.

## To Reproduce

```python
>>> a = torch.tensor([[[77.0, 24.0, 70.0, 61.0], [102.0, -16.0, 120.0, 86.0], [100.0, 52.0, -10.0, 30.0], [25.0, -90.0, -55.0, -40.0]], [[167.0, -109.0, 141.0, 198.0], [4.0, -14.0, -10.0, 9.0], [14.0, 26.0, -105.0, 9.0], [10.0, -14.0, -96.0, 18.0]]], dtype=torch.float64)
# CPU results match for batched and unbatched cases:
>>> torch.eq(torch.lu(a)[0], torch.stack([torch.lu(a[0])[0], torch.lu(a[1])[0]]))
tensor([[[True, True, True, True],
         [True, True, True, True],
         [True, True, True, True],
         [True, True, True, True]],

        [[True, True, True, True],
         [True, True, True, True],
         [True, True, True, True],
         [True, True, True, True]]])
# But CUDA results differ:
>>> a = a.cuda()
>>> torch.eq(torch.lu(a)[0], torch.stack([torch.lu(a[0])[0], torch.lu(a[1])[0]]))
tensor([[[ True,  True,  True,  True],
         [ True,  True,  True,  True],
         [ True,  True, False,  True],
         [ True,  True,  True, False]],

        [[ True,  True,  True,  True],
         [ True,  True, False, False],
         [ True,  True,  True,  True],
         [ True,  True,  True, False]]], device='cuda:0')
```

## Expected behavior

I expected that the batched and unbatched cases would give the same results.

## Possible fix

This seems to be caused by a difference in results returned by the batched/unbatched versions of the MAGMA functions `magma_*getrf_batched` and `magma_*getrf_gpu`. Perhaps we could always use the batched version in all cases. In my local source tree, I removed the special case for batch size 1 in the code (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/BatchLinearAlgebra.cu#L1772), and the problem went away. Although I don't know what the performance implications of this solution would be. I can submit a PR with this fix if it seems like an acceptable solution.

## Environment

PyTorch version: 1.9.0a0+git92770d2
Is debug build: False
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (crosstool-NG 1.24.0.133_b0863d8_dirty) 9.3.0
Clang version: Could not collect
CMake version: version 3.19.6

Python version: 3.9 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 11.0.221
GPU models and configuration: 
GPU 0: TITAN RTX
GPU 1: TITAN RTX

Nvidia driver version: 460.27.04
cuDNN version: /usr/local/cuda-10.2.89/targets/x86_64-linux/lib/libcudnn.so.7
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.1
[pip3] torch==1.9.0a0+gitf8840fa
[conda] magma-cuda110             2.5.2                         1    pytorch
[conda] mkl                       2020.4             h726a3e6_304    conda-forge
[conda] mkl-include               2020.4             h726a3e6_304    conda-forge
[conda] numpy                     1.20.1           py39hdbf815f_0    conda-forge
[conda] torch                     1.9.0a0+gitf8840fa           dev_0    <develop>

## Additional context

This is causing some issues for me in #54787 because the result of `lu` is used to determine if a matrix is singular or nonsingular, and to decide which of two different derivative formulas should be used for `det`'s backward function: https://github.com/pytorch/pytorch/pull/54787#issuecomment-821608667

cc @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @walterddr @IvanYashchuk @xwang233 @ngimel

**Comment by IvanYashchuk** [comment]
CPU results match for batched and single input cases because the CPU path for batched inputs is implemented using a for loop calling the same function. In the case of CUDA inputs, we can't expect element-wise equality because two different MAGMA functions are used for batched and non-batched inputs, they compute only approximately the same result.

**Comment by kurtamohler** [comment]
@IvanYashchuk, right that makes sense. As I mentioned in the issue description, it's possible to change `lu` to always call the batched MAGMA function to fix this inconsistency. Although I don't know how different the performance would be

**Comment by IvanYashchuk** [comment]
The performance would be different. The batched MAGMA function is not designed for larger sizes, see [the warning in the source code](https://bitbucket.org/icl/magma/src/bd42578f18a06462d7510c59b0dcc4e8f4b8375c/src/zgetrf_batched.cpp#lines-125).
We are also in the process of introducing cuBLAS (for batched) & cuSOLVER (for single input) backends for CUDA versions 10 and above (ref. https://github.com/pytorch/pytorch/issues/47953). It would bring another inconsistency.

**Comment by mruberry** [comment]
Thanks for filing this issue, @kurtamohler. I propose we address this by documenting the discrepancy when describing the implementation of torch.lu. 

cc @Lezcano

**Comment by Lezcano** [comment]
@kurtamohler About consistency of linalg functions (those in `torch.linalg` + other like `torch.lu` and so on). That is terribly annoying, indeed. A quick search on the internet says that the derivative of the determinant is zero if the n x n matrix has rank strictly less than n-1. Now the question would be, is the derivative of `det` continuous? I want to say that it is not by your comment saying that "[choosing one formula over other] gives wildly different results". It's the only explanation if the determinant only varies by 1e-10 and the gradients are very different.

If this is the case, then we would be in a similar scenario as that happening in matrix factorisations (qr, svd, eig, eigh...). In that case, there's not much we can do but to document it, as @mruberry suggests...

**Comment by kurtamohler** [comment]
@Lezcano, I also get the impression that the derivative of `det` is not continuous. However, I'm not too sure about the math for the singular matrix case. @nikitaved suggested I assume
```
u, s, v = a.svd()
det(a) = det(u) * prod(s) * det(v)
```

and derive a formula for the gradient of `det(a)` wrt `a` from that. I'd appreciate your perspective on that idea. Maybe we should continue this discussion on #54787

**Comment by Lezcano** [comment]
Closing, as there's not much that can be done. We often call slightly less accurate algorithms in the batched case to favour speed. This is a pattern that we follow all throughout the linear algebra part of PyTorch.
